### PR TITLE
Queues- Danielle Birbal - Ride-Share-Two

### PR DESCRIPTION
**Pull request by birbalds**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |   Using the Trip class and its methods in the Driver and Rider classes to grab information was a design decision I made. I considered calling on the Driver class from the Rider class but to keep the program less tangled I stuck with only Trip.     |
| Describe a concept that you gained more clarity on as you worked on this assignment. | I further clarified the difference between CSV.open versus CSV.read   |
| Describe a _nominal test_ that you wrote for this assignment. |  My tests that verify the accuracy of the data upon initialization are nominal tests.    |
| Describe an _edge case test_ that you wrote for this assignment. |  One edge case I wrote tests for the first and last trips and whether data is accurately manipulated for them. |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? |  I did poorly because my pseudocode quickly turned into actual code as I wrote it.  |

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+        attr_accessor :driver_id, :name, :vin
+
+        def initialize(id)
+            raise ArgumentError, "#{id} is not a valid ID number, only positive whole numbers are accepted" unless id.is_a?(Integer) && id >= 1
+            raise ArgumentError, "#{id} Cannot be found" unless (DRIVER_INFO.map { |line| line[0].to_i }).include? id
+
+            DRIVER_INFO.each do |line|
+                next unless line[0].to_i == id
+                @driver_id = id
```
Comment (droberts-sea) [inline lib/driver.rb:15]: I like the idea here - being able to initialize a driver with just the ID and have the program fill in the details is very convenient. However, I'm a little worried about the efficiency of looping through the whole `DRIVER_INFO` array here. In particular, if `Driver.all` creates a new `Driver` for every ID in `DRIVER_INFO`, then `Driver.all` ends up being `O(n^2)`.

It might be you could chose a different data structure to get around this. I'll leave that up to you.

File: lib/driver.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+            raise ArgumentError, "#{vin} is an invalid VIN number" unless @vin.length == 17
+        end
+
+        def trips
+            Trip.driver_trips(@driver_id)
+        end
+
+        def avg_rating
+            @driver_trips = trips
```
Comment (droberts-sea) [inline lib/driver.rb:28]: I'm curious why you chose to use an instance variable here. Seems like the trips don't need to persist long term, only for the duration of the method, which to me indicates that a local variable might be a better choice.

File: specs/driver_spec.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    describe 'avg_rating' do
+        it 'Returns a Float' do
+            new_driver.avg_rating.must_be_kind_of Float
+        end
+        it 'Cant have a negative rating' do
+            new_driver.avg_rating.wont_be :<, 0
+        end
+        it 'Calculates the average rating correctly' do
+            new_driver.avg_rating.must_equal 3.4
+        end
```
Comment (droberts-sea) [inline specs/driver_spec.rb:39]: What happens if the driver hasn't made any trips yet?

File: specs/driver_spec.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
+require_relative 'spec_helper'
+
+describe RideShare::Driver do
+    let(:new_driver) { RideShare::Driver.new(18) }
+
+    describe 'initialize' do
```
Comment (droberts-sea) [inline specs/driver_spec.rb:6]: Good organization - the nested `describes` make this code much easier to read.

File: specs/rider_spec.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            proc { RideShare::Rider.new(8_387_764) }.must_raise ArgumentError
+            proc { RideShare::Rider.new(0) }.must_raise ArgumentError
+        end
+    end
+    describe 'previous_drivers' do
+        it 'Returns an array of accurate driver data' do
+            new_rider.previous_drivers[0].driver_id.must_equal 69
+            new_rider.previous_drivers[0].name.must_equal 'Ernesto Torp'
+            new_rider.previous_drivers[0].vin.must_equal 'RF4BPA803R4AACTR1'
+        end
```
Comment (droberts-sea) [inline specs/rider_spec.rb:32]: As with `Driver.average_rating`, what happens if the rider hasn't taken any trips?

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
|  **Baseline** |   |
|   Used Git Regularly	|   You've made a good start here, but I'd like even more details. Rather than "Added Driver and Trip functionality", something like "Implemented Driver#average_raiting" would be good. That might mean you have to break your work down into smaller commits.	|
|   Answer comprehension questions	|   Regarding the last question: on class projects like these, it often feels like it's not worth spending the time pseudocoding or diagramming when you can see the solution and just go right to it. However, these skills will be invaluable when you get out into the real world and are faced with really difficult problems. It's definitely worth your while to slow down and practice them now, when the problem scope is relatively small. |
|  **Driver** |  see inline comments |
|   Uses the `all` method in the `find` method | yes |
|   Has appropriate edge-case tests for each method in the class |   missing one - see inline comment |
|   Created a method that uses a method from the `Trip` object to retrieve the list of trips | yes |
|   Created a method that uses the internal trips list to calculate the average rating | yes |
|  **Rider** |   |
|   Uses the `all` method in the `find` method | yes |
|   Has appropriate edge-case tests for each method in the class |   mostly - see inline |
|   Created a method that uses a method from the `Trip` object to retrieve the list of trips | yes |
|   Created a method that uses the internal trips method to retrieve the associated drivers | yes |
|  **Trip** |   |
|   Reads the CSV file in the `all` method | yes |
|   Has appropriate edge-case tests for each method in the class |   yes |
|   Created a method that uses a method from the `Driver` to retrieve the associated driver instance  |   yes |
|   Created a method that uses a method from the `Rider` to retrieve the associated rider instance |   yes |
|   Created a method to retrieve all trips by driver id |    yes |
|   Created a method to retrieve all trips by rider id |     yes |
|   **Overall**	|   	|

Great job! Code is clear and readable, and tests cover almost everything I was looking for. Keep up the good work!
